File: src/SemanticTypeForgers/forger.java
package SemanticTypeForgers;

import java.util.Map;
import java.util.regex.Pattern;
import java.util.Set;
import java.util.HashSet;
import java.util.List;
import java.util.Date;
import java.util.Collection;

public class forger {

    public static class AtomicBehavior<T> {
        private String name;

        public AtomicBehavior(String name) {
            this.name = name;
        }

        public T setSemanticType(T v) {
            return v;
        }

        public T getPrimitiveType(T v) {
            return v;
        }

        public boolean validate(Object value) {
            if ("PersonEmail".equals(this.name)) {
                Object prim = convertToPrimitive(value);
                if (prim instanceof String) {
                    return Pattern.matches("^[a-zA-Z0-9_.+\\-]+@[a-zA-Z0-9\\-]+\\.[a-zA-Z0-9\\-.]+$", (String) prim);
                }
                return false;
            }
            return false;
        }

        private Object primitiveStringToValue(String val) {
            String trimmed = val.trim();
            String lower = trimmed.toLowerCase();

            if (trimmed.isEmpty()) return val;
            if (val.equals("{}") || val.equals("[]")) return 0;
            if (lower.equals("true")) return true;
            if (lower.equals("false")) return false;
            if (lower.equals("null") || lower.equals("undefined")) return null;

            Pattern pattern = Pattern.compile("^[+-]?(?:(?:\\d+\\.?\\d*)|(?:\\.\\d+))(?:e[+-]?\\d+)?$", Pattern.CASE_INSENSITIVE);
            if (pattern.matcher(trimmed).matches()) {
                try {
                    if (trimmed.contains(".") || lower.contains("e")) {
                        return Double.parseDouble(trimmed);
                    }
                    return Long.parseLong(trimmed);
                } catch (NumberFormatException e) {
                    // Fallthrough
                }
            }

            return val;
        }

        private Object unwrap(Object val, Set<Integer> seen, int depth) throws Exception {
            if (depth > 100) {
                throw new Exception("convertToPrimitive: recursive unwrap limit exceeded");
            }

            if (val == null) return null;

            if (val instanceof Number || val instanceof Boolean) {
                return val;
            }

            if (val instanceof String) {
                return primitiveStringToValue((String) val);
            }

            if (val instanceof Collection) {
                Collection<?> c = (Collection<?>) val;
                if (c.isEmpty()) return null;
                return unwrap(c.iterator().next(), seen, depth + 1);
            }

            if (val instanceof Map) {
                Map<?, ?> map = (Map<?, ?>) val;
                if (map.isEmpty()) return 0;
                return unwrap(map.values().iterator().next(), seen, depth + 1);
            }

            if (val instanceof Date) {
                return ((Date) val).getTime();
            }

            // Primitive arrays, objects, circular reference detection
            int id = System.identityHashCode(val);
            if (seen.contains(id)) {
                throw new Exception("convertToPrimitive: circular reference detected");
            }
            seen.add(id);

            // Assume the object doesn't have a getPrimitiveType in simple Java translation unless explicitly interfaced.
            // Simplified return null for unknown complex objects to match JS loosely
            return null;
        }

        public Object convertToPrimitive(Object value) {
            try {
                return unwrap(value, new HashSet<>(), 0);
            } catch (Exception e) {
                return null;
            }
        }

        public T forge(T v) {
            if ("PersonEmail".equals(this.name) && !validate(v)) {
                throw new RuntimeException("Validation failed for SemanticType: " + this.name);
            }
            return v;
        }

        @SuppressWarnings("unchecked")
        public T proccessValue(Object value) {
            if (validate(value)) {
                return (T) value;
            }

            if (value != null && value instanceof Map) {
                Map<String, Object> obj = (Map<String, Object>) value;
                Object pPrice = convertToPrimitive(obj.get("productPrice"));
                Object pDiscount = convertToPrimitive(obj.get("productDiscount"));
                Object dPrice = convertToPrimitive(obj.get("deliveryPrice"));
                Object pFees = convertToPrimitive(obj.get("paymentFees"));

                double discount = (pDiscount instanceof Number) ? ((Number) pDiscount).doubleValue() : 0.0;
                double delivery = (dPrice instanceof Number) ? ((Number) dPrice).doubleValue() : 0.0;
                double fees = (pFees instanceof Number) ? ((Number) pFees).doubleValue() : 0.0;

                if (pPrice instanceof Number) {
                    double price = ((Number) pPrice).doubleValue();
                    double finalPrice = price - discount + delivery + fees;
                    return forge((T) Double.valueOf(finalPrice));
                }
            }

            Object primitiveVal = convertToPrimitive(value);
            return forge((T) primitiveVal);
        }
    }
}

File: src/SemanticTypeForgers/forger.ex
defmodule SemanticTypeForgers.Forger do
  defmodule AtomicBehavior do
    def new(name) do
      %{name: name}
    end

    def set_semantic_type(v), do: v
    def get_primitive_type(v), do: v
    def validate(behavior, value) do
      if behavior.name == "PersonEmail" do
        prim = convert_to_primitive(value)
        if is_binary(prim) do
          Regex.match?(~r/^[a-zA-Z0-9_.+\-]+@[a-zA-Z0-9\-]+\.[a-zA-Z0-9\-.]+$/, prim)
        else
          false
        end
      else
        false
      end
    end

    def primitive_string_to_value(val) do
      trimmed = String.trim(val)
      lower = String.downcase(trimmed)

      cond do
        trimmed == "" -> val
        val == "{}" or val == "[]" -> 0
        lower == "true" -> true
        lower == "false" -> false
        lower == "null" or lower == "undefined" -> nil
        true ->
          if Regex.match?(~r/^[+-]?(?:(?:\d+\.?\d*)|(?:\.\d+))(?:e[+-]?\d+)?$/i, trimmed) do
            if String.contains?(trimmed, ".") or String.contains?(lower, "e") do
              {f, _} = Float.parse(trimmed)
              f
            else
              {i, _} = Integer.parse(trimmed)
              i
            end
          else
            val
          end
      end
    end

    def unwrap(val, _seen \\ MapSet.new(), depth \\ 0) do
      if depth > 100 do
        raise "convertToPrimitive: recursive unwrap limit exceeded"
      end

      cond do
        is_nil(val) -> nil
        is_integer(val) or is_float(val) or is_boolean(val) -> val
        is_binary(val) -> primitive_string_to_value(val)
        is_list(val) ->
          if length(val) == 0, do: nil, else: unwrap(hd(val), _seen, depth + 1)
        is_map(val) ->
          if map_size(val) == 0 do
            0
          else
            [first | _] = Map.values(val)
            unwrap(first, _seen, depth + 1)
          end
        true -> nil
      end
    end

    def convert_to_primitive(value) do
      try do
        unwrap(value)
      rescue
        _ -> nil
      end
    end

    def forge(behavior, v) do
      if behavior.name == "PersonEmail" and not validate(behavior, v) do
        raise "Validation failed for SemanticType: #{behavior.name}"
      end
      v
    end

    def proccess_value(behavior, value) do
      if validate(behavior, value) do
        value
      else
        if is_map(value) and map_size(value) > 0 and (Map.has_key?(value, "productPrice") or Map.has_key?(value, :productPrice)) do
          p_price = convert_to_primitive(Map.get(value, "productPrice") || Map.get(value, :productPrice))
          p_discount = convert_to_primitive(Map.get(value, "productDiscount") || Map.get(value, :productDiscount)) || 0
          d_price = convert_to_primitive(Map.get(value, "deliveryPrice") || Map.get(value, :deliveryPrice)) || 0
          p_fees = convert_to_primitive(Map.get(value, "paymentFees") || Map.get(value, :paymentFees)) || 0

          if is_number(p_price) do
            forge(behavior, p_price - p_discount + d_price + p_fees)
          else
            forge(behavior, convert_to_primitive(value))
          end
        else
          forge(behavior, convert_to_primitive(value))
        end
      end
    end
  end
end
